### DISCUSSION TITLE
Generate PDFs in browser using Quarto, Typst and OJS

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#7633] [Q&A] [answered]
### Description

Not sure if this is feasible, but I am trying to take user input and generate PDFs in the browser using Quarto, Typst and OJS without R or Python.

For a dummy example, see [this repo](https://github.com/royfrancis/quarto-typst-ojs/). I have a directory structure that looks like this.

```
.
├── _extensions/
├── dept-news.qmd
├── input.txt
├── newsletter-cover.jpg
├── index.html
├── index.qmd
├── index_files/
├── pdfs/
│   ├── Green Award.pdf
│   └── Sixtus Award.pdf
└── script.sh
```

The `_extensions` folder is a typst format. `dept-news.qmd` uses the typst template which can be converted to pdf using quarto. This qmd contains a metadata parameter called `award`. The `input.txt` has different arguments for the `award` parameter. The `script.sh` reads arguments from `input.txt` and renders pdfs which is collected into a folder `pdfs`. This all works so far.

Now, I don't want the `input.txt`. I want that to come from a user input in `index.qmd` which runs ojs chunks. Is it possible to run the equivalent of `script.sh` in an ojs chunk? ie; read arguments and then run quarto render on `dept-news.qmd`.

Is this going to work?

`quarto 1.4.506`

**Comment by mcanouil** [comment, marked as answer]
Quarto is a command line tool.
It is not compiled in web assembly thus cannot work in the browser.
You need a server to run Quarto with or without computations via Jupyter or knitr.

Short answer it is not going to work and web assembly compilation will not happen anytime soon.

**Comment by cderv** [comment]
> Is it possible to run the equivalent of script.sh in an ojs chunk? ie; read arguments and then run quarto render on dept-news.qmd.

I would look into making a webapp for what you want to do. You can easily make one using Shiny with R or Python - this web app would take your user input, and trigger the rendering on Quarto on the server where it is hosted, and then show the preview of the rendered document. 

As shared in other answer, you need a server to run quarto.

**Reply by royfrancis** [reply]
Thanks Christophe! Yes, I currently use a shiny app solution, but adding R and shiny increases the docker container size a lot. So I was just wondering if there can be a light weight solution that excludes R, Python and Shiny altogether. I was intrigued by OJS since it has widgets and takes in user input. But quarto render in the browser seems to be problem.

This question can be rephrased to what is a light weight GUI that can be added to Quarto? Perhaps there is a Lua solution.

**Reply by cderv** [reply]
> This question can be rephrased to what is a light weight GUI that can be added to Quarto? Perhaps there is a Lua solution.

The problem is not about the GUI, it is about the fact that Quarto needs to be run on a server. So you need a server part, and not just a GUI part. 

From there, you can build any solution that will pass information to a server - it could be a web app using a framework that already handles passing information from / to server (like shiny), or you could also use an API on your server (using R or Python, or node.js or anything to your liking), that would allow small UI to call the API and get back some information. 

You can definitely build something like this in node JS for example if you don't want R or python.
